### PR TITLE
fix: Change AggregateRoot's commit() to use publishAll()

### PR DESCRIPTION
**Pull request by goldsam**
Using publishAll() instead of publish() enables implementations of IEventBus to
support correct transactional semantics when used together with EventPublisher.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?
`AggregateRoot`'s current `commit()` implementation invokes `publish()` instead of a `publishAll()`, This prevents `IEventPublisher.publishAll()` from ever being called which breaks transactional semantics when publishing events to an `IEventPublisher` implementation. 

Issue Number: #325

## What is the new behavior?

A new `publishAll()` method has been introduced to `AggregateRoot` and `EventPublisher` has been updated to use the corresponding API of `IEventBus`. This allows the user to implement the already existing `IEventPublisher.publishAll()` method to support atomic/transactional event publishing. 

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```
I don't believe this introduces any breaking changes due to the following two observations:
* The default implementation of `AggregateRoot.publish()` and the new `AggregateRoot.publishAll()` both do nothing.
* The overridden implementation of the new `AggregateRoot.publishAll()`, when used together with `EventPublisher`, will invoke `EventBus.publishAll()`. The current implementation of `EventBus.publishAll()` assumes the original behavior (that is, invoking `IEventPublisher.publish()` for each event) unless the user has explicitly implemented `IEventPublisher.publishAll()` (likely with intent to support an atomic commit).

## Other information

none

### CODE REVIEW

File: src/event-publisher.ts [hunk 1]
```diff
@@ -16,16 +16,16 @@ export class EventPublisher<EventBase extends IEvent = IEvent> {
   ): T {
     const eventBus = this.eventBus;
     return class extends metatype {
-      publish(event: EventBase) {
-        eventBus.publish(event);
+      publishAll(events: EventBase[]) {
+        eventBus.publishAll(events);
```
Comment (kamilmysliwiec) [inline src/event-publisher.ts:20]: Shouldn't we leave the `publish()` implementation and just add the `publishAll()`?

File: src/event-publisher.ts [hunk 2]
```diff
@@ -16,16 +16,16 @@ export class EventPublisher<EventBase extends IEvent = IEvent> {
...
       }
     };
   }
 
   mergeObjectContext<T extends AggregateRoot<EventBase>>(object: T): T {
     const eventBus = this.eventBus;
-    object.publish = (event: EventBase) => {
-      eventBus.publish(event);
+    object.publishAll = (events: EventBase[]) => {
+      eventBus.publishAll(events);
```
Comment (kamilmysliwiec) [inline src/event-publisher.ts:28]: same as above

File: src/event-publisher.ts (hunk 1 above)
Comment (goldsam) [reply src/event-publisher.ts:20]: Yes, I suppose that does make a bit more sense.

File: src/event-publisher.ts (hunk 2 above)
Comment (goldsam) [reply src/event-publisher.ts:28]: Agreed.

**Comment by goldsam** [comment]
I made an additional update yesterday with your requested changes.

**Comment by goldsam** [comment]
I was pretty responsive in making your requested changes. Its been quite some time now. Are you going to accept my PR?

**Comment by kamilmysliwiec** [comment]
As you can see, this PR has the `blocked` status which means that this change is actually introducing a breaking change. To merge this, we would have to publish a new major release (v8) which isn't possible yet (we need to wait till the next major release of the framework and the whole ecosystem)

**Comment by goldsam** [comment]
> ## Does this PR introduce a breaking change?
>```
>[ ] Yes
>[x] No
>```

I don't mean to sound irritated or harsh, but I've been left watching this issue without any feedback. In the future, it would be helpful if some feedback was provided about your decision when updating an issue's status or labels. Upon further review, I acknowledge that this would be a breaking change. I look forward to future release. Thank you for your work.

**Comment by underfisk** [comment]
This feature is very useful for parallel execution, one thing that it lacks is executed in parallel, if you take a look at the source it gets within a loop (forEach) dispatched 1 by 1 but if the AggregateRoot has 10 pending events at the commit phase, what will happen then? They will not be handled in parallel and the effect of having something like Promise.all is very handy for transactions
@goldsam I agree with your points, I don't consider it a breaking change but an enhancement
@kamilmysliwiec If you can take a look at this performance consideration, that would be great to be seen in another release

**Comment by kodeine** [comment]
Any ideas when is this going to be merged?

**Comment by underfisk** [comment]
@jmcdo29 We talked about this feature in discord, do you think there is any chance at least of obtaining some feedback from the team regarding this change? This batching is way important and can improve performance in some cases

**Pull request merged**